rl4co/models/zoo/common/autoregressive/decoder.py

```diff
@@ -133,74 +136,38 @@ def forward(
                 - "greedy": take the argmax of the logits
                 - "multistart_sampling": sample as sampling, but with multi-start decoding
                 - "multistart_greedy": sample as greedy, but with multi-start decoding
-            num_starts: Number of multi-starts to use. If None, will be calculated from the action mask
+                - "beam_search": perform beam search
             softmax_temp: Temperature for the softmax. If None, default softmax is used from the `LogitAttention` module
             calc_reward: Whether to calculate the reward for the decoded sequence
+            strategy_kwargs: Keyword arguments for the decoding strategy. See :class:`rl4co.models.nn.dec_strategies.DecodingStrategy`
 
         Returns:
             outputs: Tensor of shape (batch_size, seq_len, num_nodes) containing the logits
             actions: Tensor of shape (batch_size, seq_len) containing the sampled actions
             td: TensorDict containing the environment state after decoding
         """
-
         # Instantiate environment if needed
         if isinstance(env, str):
             env_name = self.env_name if env is None else env
             env = get_env(env_name)
 
-        # Multi-start decoding. If num_starts is None, we use the number of actions in the action mask
-        if "multistart" in decode_type:
-            if num_starts is None:
-                num_starts = get_num_starts(td, env.name)
-        else:
-            if num_starts is not None:
-                if num_starts > 1:
-                    log.warn(
-                        f"num_starts={num_starts} is ignored for decode_type={decode_type}"
-                    )
-            num_starts = 0
-
         # Compute keys, values for the glimpse and keys for the logits once as they can be reused in every step
         cached_embeds = self._precompute_cache(embeddings, td=td)
 
-        # Collect outputs
-        outputs = []
-        actions = []
-
-        # Multi-start decoding: first action is chosen by ad-hoc node selection
-        if num_starts > 1 or "multistart" in decode_type:
-            action = self.select_start_nodes_fn(td, env, num_starts=num_starts)
-
-            # Expand td to batch_size * num_starts
-            td = batchify(td, num_starts)
-
-            td.set("action", action)
-            td = env.step(td)["next"]
-            log_p = torch.zeros_like(
-                td["action_mask"], device=td.device
-            )  # first log_p is 0, so p = log_p.exp() = 1
-
-            outputs.append(log_p)
-            actions.append(action)
+        # setup decoding strategy
+        self.decode_strategy: DecodingStrategy = get_decoding_strategy(
```

Comment:
I wonder if instantiating the strategy each time has some impact on speed? I guess not much. In case there is indeed some difference, it might be worth it to "cache" the strategy - i.e. since we are saving it in `self`, then we could do something like:

```python

if self.decode_strategy.name != decode_type:
  self.decode_strategy: DecodingStrategy = get_decoding_strategy(
              decode_type, **strategy_kwargs
          )
```

Reply:
Thanks for the quick review @fedebotu! Great catch, making the DecodingStrategy a subclass of nn.Module is indeed not necessary. After removing that, I did some profiling to check the impact of instantiating the strategy each time. Results indicate a super minor impact on speed, please see the image below. For the test, I implemented an optional cache, similar to what you proposed, and checked the speed with and without caching the strategy. 
From a readability point-of-view, I would argue that the cache is not necessary. What do you suggest? :)

![Bildschirmfoto 2024-01-29 um 13 02 20](https://github.com/ai4co/rl4co/assets/66203600/c8824539-f737-4ea1-ad1f-bdf82a416353)

Reply:
Now that the class instantiation is simple, there is basically no overhead and much cleaner code. Let's keep it as is!